sos/collector/clusters/saltstack.py

```diff
@@ -0,0 +1,107 @@
+# Copyright Red Hat 2022, Trevor Benson <trevor.benson@gmail.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import json
+from shlex import quote
+from sos.collector.clusters import Cluster
+
+
+class saltstack(Cluster):
+    """
+    The saltstack cluster profile is intended to be used on saltstack
+    clusters (Salt Project).
+    """
+
+    cluster_name = "Saltstack"
+    packages = ("salt-master",)
+    sos_plugins = ["saltmaster"]
+    cmd = "salt-run --out=pprint manage.up"
+    hostname_cmd = "salt --out=newline_values_only {minion} grains.get fqdn"
+    strict_node_list = True
+    option_list = [
+        ("all_nodes", False, "Filter node list to all nodes, even if "
+         " salt-mionion service is down."),
+        ("compound", "", "Filter node list to those matching compound"),
+        ("glob", "", "Filter node list to those matching glob pattern"),
+        ("grain", "", "Filter node list to those with matching grain"),
+        ("hostnames", False, "When minion isn't resolvable for control_persist"
+         " this will return the fqdn of the minion. This option is not"
+         " compatible with option all_nodes or transport saltstack"),
+        ("list", "", "Filter node list to those matching list"),
+        ("nodegroup", "", "Filter node list to those matching nodegroup"),
+        ("pillar", "", "Filter node list to those with matching pillar"),
+        ("regex", "", "Filter node list to those matching regex"),
+        ("subnet", "", "Filter node list to those in subnet"),
+    ]
```

Comment:
A few thoughts here...

- The default behavior of a cluster profile should be to return all nodes in the cluster, and from there various options should allow reducing/filtering it.
- The `list`, `glob`, and `regex` options are redundant with the current behavior of `--nodes`.
- For `hostnames`, I'm confused. If we can't resolve the host, why return an FQDN that needs to be resolvable in order to be useful?

Reply:
> * The default behavior of a cluster profile should be to return all nodes in the cluster, and from there various options should allow reducing/filtering it.

I'll flip the behavior then and make it return every minion, even the offline ones, and make the option for "up_nodes" which are online. This leads to salt errors when a targeted minion does not respond, which is why the default behavior started off as only online nodes.

> * The list, glob, and regex options are redundant with the current behavior of --nodes.

For what transports? The saltstack transport currently only works with `--nodes` in a list format. Admittedly this is probably due to only coding it to use a list format when I wrote the transport. I don't recall noticing the  RemoteTransport having any specific handoff of nodes as one type vs the other and I did not write a parser to determine it in the saltstack transport.

I do see that `--help` mentions nodes can be a regex, but I'm not having any luck with testing that against **control_persist** transport. **Test Environment**: 6 hosts named **store-1** through **store-6**. GLOBS: Using `--nodes store-*`, `--nodes store-\*`, `--nodes 'store-*'`, and REGEX: `--nodes 'store-(1|2)'` where none of these return any nodes in the list to collect from, except the localhost. Unless I'm missing the requirements for it? These exist in my `/etc/hosts`, but are not defined inside `~/.ssh/config` if that is related somehow

> * For hostnames, I'm confused. If we can't resolve the host, why return an FQDN that needs to be resolvable in order to be useful?

Saltstack works via minion ID's unless the administrator makes the minion ID match the hosts shortname or fqdn. This is sometimes the case but with no requirement it often is not. Advanced targeting allows using grains, which then could target based on a hostname value.

I presumed a collector cluster type of saltstack would mostly likely used with the collector transport of saltstack. Therefore I returned the minion ID since that would be what saltstack definitely uses. If the administrator does not use the FQDN as the minion ID, then using it as the default returned value would not work for saltstack transport, but would for others. 

I realize during this conversation I have access to the `self.opts.transport` attribute and could abstract this hostnames tranport option entirely. I will probably adjust this to use if/else `if self.opts.transport == "saltstack"`. **If** the transport is saltstack return the nodes as minion IDs. **Else** parse each node (by minion ID) for it's `fqdn` "grain" and append it to the list of nodes, returning only resolvable (presumably) hostnames.

Reply:
To clarify, this is how `--nodes` operates:

First, we get all the nodes from the cluster that the profile decides to return. From there, any regex entries for `nodes` acts as a filter - so using `--nodes store-(1|2)` would only allow the nodes `store-1` and `store-2` **if** the cluster profile returned them. Of important note here is that this uses python regex, not shell-style globs. So `store-*` would not return what you're probably looking for, but `store-.*` would.

Finally, any non-regex entries for `--nodes` that look like regular hostnames are added on top of any nodes returned by the cluster profile.

To be clear, you cannot use a regex to **add** nodes to the list, since we have no way of validating what that list should expand to - we only use regexes as a filter to reduce the node list.

I've done some brief testing locally and `--nodes` is working as described as far as I can tell, however if there's some condition that may be influencing that, let's investigate and fix that.

I could maybe be persuaded about the `glob` option here, but `list` and `regex` are absolutely redundant of the base `--nodes` option, and I think we should err on the side of a consistent user experience between cluster profiles for these options.

Reply:
@TurboTurtle 

Your testing of control_persist using a regex "like" `store-(1|2)` worked? As I mentioned this did **not** work in my test when you first mentioned this.

In my test the only node was the localhost doing the collect command:
* Using it bare `--nodes store-(1|2)` causes `-bash: syntax error near unexpected token `('`.
* Using it quoted `--nodes 'store-(1|2)'` results in finding ONLY the localhost:
```
The following is a list of nodes to collect from:
	EL8_supervisor

Press ENTER to continue with these nodes, or press CTRL-C to quit

Connecting to nodes...
Collection would only gather from localhost due to failure to either enumerate or connect to cluster nodes. Assuming single collection from localhost is not desired.
Aborting...
```
Maybe this is a combination effect of using both a `--cluster-type X` and `--nodes` together, but neither the documentation or your earlier comment made it clear that the regex for `--nodes` is only viable when a cluster profile returns `get_nodes()`, but not when `--nodes` is used without a cluster profile?

Reply:
From `man sos-collect`:

--nodes NODES
      Provide  a comma-delimited list of nodes to collect sosreports from, or a **regex string to be used to
      compare _discovered_ node names against.** If using a regex, only nodes matching the regex will be  used
       - i.e. it can be used as a whitelist but not a blacklist.


If the cluster profile returns `store-1`, `store-2`, `store-3`, `store-4` for example, and you use `--nodes='store-(1|2)'` the list would be reduced to just `store-1` and `store-2`. If the cluster profile does **not** return `store-1` or `store-2`, this use of `--nodes` would cause **no** nodes to be returned.

Reply:
Also note, that you generally need to quote regex strings (not just for python/sos, but any tool that has you pass these) as otherwise your shell will try to interpret the characters for its own purposes.

Reply:
:+1: My oversight, I only checked the --help output originally which mentioned support for regex, but not that its a filter against the cluster profile.

Reply:
> Also note, that you generally need to quote regex strings (not just for python/sos, but any tool that has you pass these) as otherwise your shell will try to interpret the characters for its own purposes.

Yep, the first bare usage I didnt expect to work, however I wanted to be thorough in testing what your comment was showing `--nodes store=(1|2)` to be sure I did not overlook something unexpected.

Reply:
@TurboTurtle Whats the final call on the options? Remove them and force the end user to use `--nodes` if they want to use list/regex filters or leave them in place because they do not prevent the end user from using `--nodes` but would still be accessible to end users more familiar with saltstack administration?

Thanks in advance.